syntax = "proto3";

package arduino_trader.optimization;

import "common/common.proto";
import "common/position.proto";

// Optimization Service - Handles portfolio optimization
service OptimizationService {
  // Optimize portfolio allocation
  rpc OptimizeAllocation(OptimizeAllocationRequest) returns (OptimizeAllocationResponse);

  // Optimize trade execution
  rpc OptimizeExecution(OptimizeExecutionRequest) returns (OptimizeExecutionResponse);

  // Calculate optimal rebalancing
  rpc CalculateRebalancing(CalculateRebalancingRequest) returns (CalculateRebalancingResponse);

  // Health check
  rpc HealthCheck(common.Empty) returns (common.HealthCheckResponse);
}

// Request to optimize allocation
message OptimizeAllocationRequest {
  string portfolio_hash = 1;
  repeated common.Position current_positions = 2;
  repeated SecurityAllocation target_allocations = 3;
  common.Money available_cash = 4;
  OptimizationConstraints constraints = 5;
}

// Security allocation target
message SecurityAllocation {
  string isin = 1;
  string symbol = 2;
  double target_weight = 3;  // 0.0 to 1.0
  double min_weight = 4;
  double max_weight = 5;
}

// Optimization constraints
message OptimizationConstraints {
  int32 max_positions = 1;
  common.Money min_position_size = 2;
  common.Money max_position_size = 3;
  double max_turnover_pct = 4;
  bool allow_fractional = 5;
}

// Response with optimized allocation
message OptimizeAllocationResponse {
  bool success = 1;
  repeated AllocationChange changes = 2;
  double objective_value = 3;
  string solver_status = 4;
  OptimizationMetrics metrics = 5;
}

// Allocation change
message AllocationChange {
  string isin = 1;
  string symbol = 2;
  double current_weight = 3;
  double target_weight = 4;
  double quantity_change = 5;
  common.Money value_change = 6;
  common.TradeSide side = 7;
}

// Optimization metrics
message OptimizationMetrics {
  double expected_return = 1;
  double expected_volatility = 2;
  double sharpe_ratio = 3;
  double tracking_error = 4;
  int32 iterations = 5;
  double solve_time_ms = 6;
}

// Optimize execution request
message OptimizeExecutionRequest {
  string portfolio_hash = 1;
  repeated TradeOrder orders = 2;
  common.Money available_cash = 3;
}

// Trade order
message TradeOrder {
  string isin = 1;
  string symbol = 2;
  common.TradeSide side = 3;
  double quantity = 4;
  common.Money limit_price = 5;
}

// Optimize execution response
message OptimizeExecutionResponse {
  bool success = 1;
  repeated ExecutionPlan execution_plans = 2;
  common.Money total_cost = 3;
  common.Money total_impact = 4;
}

// Execution plan
message ExecutionPlan {
  string isin = 1;
  repeated ExecutionSlice slices = 2;
}

// Execution slice
message ExecutionSlice {
  double quantity = 1;
  common.Timestamp execute_at = 2;
  common.Money estimated_price = 3;
  string strategy = 4;
}

// Calculate rebalancing request
message CalculateRebalancingRequest {
  string portfolio_hash = 1;
  repeated common.Position current_positions = 2;
  repeated SecurityAllocation target_allocations = 3;
  common.Money available_cash = 4;
}

// Calculate rebalancing response
message CalculateRebalancingResponse {
  bool needs_rebalancing = 1;
  repeated AllocationChange changes = 2;
  double total_drift_pct = 3;
  common.Money estimated_cost = 4;
}
